### PR TITLE
feat(contracts): new zktrie verifier

### PR DESCRIPTION
**Pull request by zimpha**
### Purpose or design rationale of this PR

This PR implements the zktrie verifier based on latest [zktrie code](https://github.com/scroll-tech/zktrie).

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] No, this PR doesn't involve a new deployment, git tag, docker image tag

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

### CODE REVIEW

File: contracts/src/libraries/verifier/ZkTrieVerifier.sol [hunk 1]
```diff
@@ -58,19 +56,20 @@ library ZkTrieVerifier {
...
-                let success := staticcall(gas(), hasher, x, 0x44, 0x20, 0x20)
+                mstore(add(x, 0x44), domain)
+                let success := staticcall(gas(), hasher, x, 0x64, 0x20, 0x20)
                 require(success, "poseidon hash failed")
                 r := mload(0x20)
             }
             // compute poseidon hash of 1 uint256
-            function hash_uint256(hasher, v) -> r {
-                r := poseidon_hash(hasher, shr(128, v), and(v, 0xffffffffffffffffffffffffffffffff))
+            function hash_uint256(hasher, v, domain) -> r {
```
Comment (noel2004) [inline contracts/src/libraries/verifier/ZkTrieVerifier.sol]: The domain is fixed (512) while hashing a uint256 by dividing it as two 128-bit, hi/lo parts. so maybe we do not need to apply an additional parameter (`domain`) here.

File: contracts/src/libraries/verifier/ZkTrieVerifier.sol [hunk 2]
```diff
@@ -130,129 +130,121 @@ library ZkTrieVerifier {
...
-                    tmpHash2 := poseidon_hash(poseidon, tmpHash2, calldataload(ptr))
+                    tmpHash := poseidon_hash(hasher, storageRootHash, tmpHash, 1280)
+                    tmpHash := poseidon_hash(hasher, mload(0x00), tmpHash, 1280)
+                    tmpHash := poseidon_hash(hasher, tmpHash, calldataload(ptr), 1280)
                     ptr := add(ptr, 0x20) // skip PoseidonCodeHash
 
-                    tmpHash1 := poseidon_hash(poseidon, 1, key)
-                    tmpHash1 := poseidon_hash(poseidon, tmpHash1, tmpHash2)
+                    tmpHash := poseidon_hash(hasher, key, tmpHash, 4)
+                    require(eq(leafHash, tmpHash), "InvalidAccountLeafNodeHash")
```
Comment (noel2004) [inline contracts/src/libraries/verifier/ZkTrieVerifier.sol:173]: The process for hash calculation here is ok. But i advice left more comment or at least some directions (like https://github.com/scroll-tech/mpt-circuit/blob/v0.7/spec/mpt-proof.md) here to make anyone without too much backgroud knowledge are also able to have fast understanding

File: contracts/src/libraries/verifier/ZkTrieVerifier.sol [hunk 3]
```diff
@@ -58,19 +56,20 @@ library ZkTrieVerifier {
                 }
             }
             // compute poseidon hash of two uint256
-            function poseidon_hash(hasher, v0, v1) -> r {
+            function poseidon_hash(hasher, v0, v1, domain) -> r {
                 let x := mload(0x40)
-                // keccack256("poseidon(uint256[2])")
-                mstore(x, 0x29a5f2f600000000000000000000000000000000000000000000000000000000)
+                // keccack256("poseidon(uint256[2],uint256)")
+                mstore(x, 0xa717016c00000000000000000000000000000000000000000000000000000000)
```
Comment (noel2004) [inline contracts/src/libraries/verifier/ZkTrieVerifier.sol:62]: where is this magic number from? the entry of function selector?

**Review by noel2004 — DISMISSED** [review]
LGTM except for some nitpicks

File: contracts/src/libraries/verifier/ZkTrieVerifier.sol (hunk 3 above)
Comment (zimpha) [reply contracts/src/libraries/verifier/ZkTrieVerifier.sol:62]: yes

**Review by noel2004 — DISMISSED** [review]

**Review by noel2004 — APPROVED** [review]

**Review by icemelon — APPROVED** [review]

**Pull request merged**